engine/grid.py

```diff
@@ -253,44 +265,51 @@ def area(x1, y1, x2, y2, x3, y3):
 
         # Check if sum of A1, A2 and A3
         # is same as A
-        if (A == A1 + A2 + A3):
-            return True
+        return A == (A1 + A2 + A3)
+
+    def activate_line(self, row, col, n, isHorizontal):
+        """
+        Activate a single horizontal line of length n starting at row, col.
+        If isHorizontal is True, the line is from (row, col) to (row, col + n). 
+        Otherwise, the line is from (row, col) to (row + n, col). 
+        """
+        # Note: rows are y-position and columns are x-position
+        if isHorizontal:
+            for i in range(n):
+                self.nodes[col+i][row].is_active = True
         else:
-            return False
+            for i in range(n):
+                self.nodes[col][row+i].is_active = True
 
+    ##Activate triangle based on three points
     def activate_triangle(self, x1, y1, x2, y2, x3, y3):
         """
-        Activates all the nodes in a traingle based on three points.
+        Activates all the nodes in a traingle.
         """
-        rows = self.nodes.shape[0]
-        cols = self.nodes.shape[0]
-        for x in range(rows):
-            for y in range(cols):
-                if self.is_inside_triangle(x1, y1, x2, y2, x3, y3, x, y):
+        for x in range(self.num_rows):
+            for y in range(self.num_cols):
+                if self.isInsideTriangle(x1, y1, x2, y2, x3, y3, x, y):
                     self.nodes[x, y].is_active = True
 
-    
+    ##Checks if node is on border of activated nodes
     def is_on_border(self, row, col, row_limit, col_limit):
-        """
-        Checks if node is on border of activated nodes.
-        """
-        min_col = max(0, col-1)
-        min_row = max(0, row-1)
-        max_col = min(col_limit, col+1)
-        max_row = min(row_limit, row+1)
+        min_col = max(0, col - 1)
+        min_row = max(0, row - 1)
+        max_col = min(col_limit, col + 1)
+        max_row = min(row_limit, row + 1)
 
         # If this node is on the very edge of the grid, it is automatically a border node
-        if min_col == 0 or min_row == 0 or max_col == col_limit or max_row == row_limit:
+        if row == 0 or col == 0 or col == self.get_num_cols() - 1 or row == self.get_num_rows() - 1:
```

Comment:
nit, could change self.get_num_cols() -1 into var called last_col or smthg to be more descriptive about what this condition means

Reply:
I am not sure, self.get_num_cols() feels more descriptive the function does number of columns.  It is also used in other functions as well for that purpose, so I think it might be best to keep it.

Reply:
agreed w lin lets keep it